include/srf/experimental/modules/module_registry.hpp

```diff
@@ -0,0 +1,123 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <nlohmann/json.hpp>
+
+#include <functional>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <vector>
+
+namespace srf::modules {
+
+class SegmentModule;
+
+/**
+ * Simple, thread safe, global module registry.
+ */
+class ModuleRegistry
+{
+  public:
+    using module_constructor_t =
+        std::function<std::shared_ptr<srf::modules::SegmentModule>(std::string module_name, nlohmann::json config)>;
+
+    using module_registry_map_t  = std::map<std::string, module_constructor_t>;
+    using module_namespace_map_t = std::map<std::string, module_registry_map_t>;
+    using module_name_map_t      = std::map<std::string, std::vector<std::string>>;
+
+    ModuleRegistry() = delete;
+
+    /**
+     * Return true if the registry contains 'name' else false
+     * @param name Name of the module
+     * @return boolean indicating if the registry contains the required module.
+     */
+    static bool contains(const std::string& name, const std::string& registry_namespace = "default");
+
+    /**
+     * Return true if the registry contains the namespace, else false
+     * @param registry_namespace Namespace name
+     * @return boolean indicating if the registry contains the required namespace.
+     */
+    static bool contains_namespace(const std::string& registry_namespace);
+
+    /**
+     * Attempt to retrieve the module constructor for a given module name; throws an error
+     * if the given module does not exist.
+     * @param name Name of the module
+     * @return Module constructor
+     */
+    static module_constructor_t find_module(const std::string& name, const std::string& registry_namespace = "default");
+
+    /**
+     * Retrieve a map of namespace -> registered module name vectors
+     * @return Map of namespace -> registered module vector pairs
+     */
+    static const module_name_map_t& registered_modules();
+
+    /**
+     * Simple register call, places the module into the default namespace
+     * @param name Name of the module
+     * @param fn_constructor Module constructor
+     */
+    static void register_module(std::string name,
+                                const std::vector<unsigned int>& release_version,
+                                module_constructor_t fn_constructor);
+
+    /**
+     * Attempt to register the provided module constructor for the given name; throws an error
+     * if the module already exists.
+     * @param name Name of the module
+     * @param registry_namespace Namespace where the module `name` should be registered.
+     * @param fn_constructor Module constructor
+     */
+    static void register_module(std::string name,
+                                std::string registry_namespace,
```

Comment:
I see this as necessary for a good end user experience, and its been quite nice in testing done so far for segregating modules into logical bins. For instance, loading all testing modules into an 'srf_unittest' namespace.

Longer term it seems important to avoid potential naming conflicts/collisions:

Imagine two library creators, A and B, who create a shared module with 'MyModule' entry.

Assuming they register "A::MyModule" and "B::MyModule", there are no conflicts and much less chance of headaches to end users.

Reply:
But isnt the namespace up to the user of the Module? I can see the usefulness of namespacing for grouping multiple modules together. But the name collision would just be resolved by the user registering one module with one name and another module as another.

Reply:
While someone else could come in, load a module, then register it in a different namespace, in general, no the user of the Module doesn't decide on the namespace its in, they just retrieve it from a namespace.